controller/controller_server/utils.py

```diff
@@ -8,42 +10,153 @@
+    user, password, array_addresses = _get_array_connection_info_from_secret(secret)
+    return Secret(user=user, password=password, array_addresses=array_addresses, uid=secret_uid)
+
+
+def get_pool_from_parameters(parameters, secret_uid=None):
+    pools_by_system = parameters.get(config.PARAMETERS_POOLS_BY_SYSTEM)
+    if pools_by_system and secret_uid:
+        return get_pool_by_uid(pools=pools_by_system, uid=secret_uid)
+    return parameters.get(config.PARAMETERS_POOL)
+
+
+def get_volume_id(new_volume, secret_uid):
+    return _get_object_id(new_volume, secret_uid)
 
 
 def get_snapshot_id(new_snapshot):
     return _get_object_id(new_snapshot)
 
 
-def _get_object_id(obj):
+def _get_object_id(obj, secret_uid=None):
+    if secret_uid:
+        return config.PARAMETERS_OBJECT_ID_DELIMITER.join((obj.array_type, secret_uid, obj.id))
     return config.PARAMETERS_OBJECT_ID_DELIMITER.join((obj.array_type, obj.id))
 
 
-def validate_secret(secret):
-    logger.debug("validating secrets")
-    if secret:
-        if not (config.SECRET_USERNAME_PARAMETER in secret and
-                config.SECRET_PASSWORD_PARAMETER in secret and
-                config.SECRET_ARRAY_PARAMETER in secret):
+def _is_imported_string_valid(imported_string):
+    return re.match(config.SECRET_VALIDATION_REGEX, imported_string) and imported_string == imported_string.strip()
+
+
+def _validate_secret_uid(uid):
+    if not _is_imported_string_valid(uid):
+        raise ValidationException(messages.invalid_uid_value_message.format(uid))
+    if len(uid) > config.SECRET_UID_MAX_LENGTH:
+        raise ValidationException(
+            messages.parameter_length_is_too_long.format("secret_uid", uid, config.SECRET_UID_MAX_LENGTH))
+
+
+def _validate_secret(secret):
+    if not (config.SECRET_USERNAME_PARAMETER in secret and
+            config.SECRET_PASSWORD_PARAMETER in secret and
+            config.SECRET_ARRAY_PARAMETER in secret):
+        raise ValidationException(messages.invalid_secret_message)
+
+
+def _validate_topologies(secret_topologies):
+    for topologies in secret_topologies:
+        for topology_key, topology_value in topologies.items():
+            split_topology = topology_key.split(config.PARAMETERS_TOPOLOGY_DELIMITER)
+            if len(split_topology) != 2:
+                raise ValidationException(messages.invalid_topology_key_message.format(topology_key))
+            prefix, suffix = split_topology
+            if not _is_imported_string_valid(suffix):
+                raise ValidationException(messages.invalid_topology_key_message.format(suffix))
+            if prefix not in config.TOPOLOGY_PREFIXES:
+                raise ValidationException(messages.invalid_topology_key_message.format(prefix))
+            if not _is_imported_string_valid(topology_value):
+                raise ValidationException(messages.invalid_topology_value_message.format(topology_value))
```

Comment:
I think these validations should be aligned to only enforce the CSI spec, i.e. without any SP-specific validations

Reply:
what validations? what part of the spec?

Reply:
the validations in these lines, the part of the spec which states the string value constraints

Reply:
its aligned with spec

Reply:
I mean that they should also allow everything that the spec allows, e.g. not specific prefixes

Reply:
no more checking that